ls コマンドを作るオブジェクト指向版

gaki273 opened the pull request.

itagakisan committed:
実行ファイルとテストファイルを追加 機能追加

[目的]
- プログラムの実行とテストを行う。

[変更内容]
- 07.ls_objectにおいて、la.rbとls_test.rbを追加
- ls_test.rbにLsTestクラスを追加

itagakisan committed:
ls_option.rbを追加、LsOptionクラスを追加 機能追加

[目的]
- オプションの判定を行う

[変更内容]
- 07.ls_objectにおいて、ls_option.rbを追加
- ls_option.rbにLsOptionクラスを追加

itagakisan committed:
LsOptionクラスのテストを追加 テスト追加

[目的]
- LsOptionクラスの動作確認

[変更内容]
- LsTestクラスにtest_ls_option_singularメソッドとtest_ls_option_multipleメソッドを追加
- 単数オプション(-a, -r, -l)が渡された時のテストケースを追加
- 複数オプション(-ar, -al, -arl)が渡された時のテストケースを追加

itagakisan committed:
LsOptionクラスにコマンドラインオプションを解析する機能を追加 機能追加

[目的]
- コマンドラインオプションの解析ロジックをカプセル化

[変更内容]
- optparseを使用して、`-a`、`-r`、`-l`オプションを解析する機能を実装
- 解析結果を問い合わせるためのメソッド(`all?`、`reverse?`、`long_format?`)を追加

itagakisan committed:
ファイルの情報を取得するFileInfoクラスを追加 機能追加

[目的]
- ファイル情報の取得を行う。

[変更内容]
- 07.ls_objectにおいて、file_info.rbを追加
- file_info.rbにFileInfoクラスを追加

itagakisan committed:
ファイルのパーミッションを取得するPermクラスを追加 機能追加

[目的]
- ファイルのパーミッションを受け取り、出力表示への変換を行う。

[変更内容]
- 07.ls_objectにおいて、perm.rbを追加
- mode.rbにPermクラスを追加

itagakisan committed:
Permクラスのテストを追加 テスト追加

[目的]
- Permクラスの動作確認

[変更内容]
- LsTestクラスにtest_permメソッドを追加
- ディレクトリと通常ファイルのテストケースを追加

itagakisan committed:
Permクラスに定数を追加 機能追加

[目的]
- 対応したファイルタイプとパーミッションの文字列に変換を行う

[変更内容]
- ファイルタイプとパーミッションの対応を定義した`FILE_TYPE`と`FILE_PERMISSION`定数を追加

itagakisan committed:
Permクラスに`trans_type_and_permission`メソッドを追加 機能追加

[目的]
- 対応したファイルタイプとパーミッションの文字列に変換する処理を実装

[変更内容]
- `trans_type_and_permission`メソッドを追加
- `File::Stat`オブジェクトを受け取り、定義済みの定数を用いてファイルタイプとパーミッション文字列に変換する

itagakisan committed:
FileInfoクラスのテストを追加 テスト追加

[目的]
- FileInfoクラスの動作確認

[変更内容]
- LsTestクラスに`test_file_info`メソッドを追加
- FileInfoクラスの各メソッド（`file_mode`, `link`, `owner`, `group`, `bytesize`, `update_time`, `file_type`）の返り値を確認するテストを追加

itagakisan committed:
FileInfoクラスに各情報を取得するメソッドを追加 機能追加

[目的]
- ファイルの情報を取得する

[変更内容]
- `initialize`でファイル名を受け取り、`File::Stat`オブジェクトをインスタンス変数に保持
- モード、リンク数、オーナー、グループ、サイズ、更新日時などの個別の情報を返すpublicメソッドを実装
- `-l`オプションが有効な場合にシンボリックリンクの表示を整形する`file_type`メソッドを実装

itagakisan committed:
FileInfoクラスのブロック数を取得するメソッドのテストを追加 テスト追加

[目的]
- ブロック数を取得する機能の動作確認

[変更内容]
- `test_file_info`メソッドに、`blocks`メソッドの返り値を確認するテストケースを追加

itagakisan committed:
FileInfoクラスのブロック数を取得するメソッドを追加 機能追加

[目的]
- ブロック数を取得する機能の実装

[変更内容]
- `FileInfo`クラスに、ファイルのブロック数を計算して返す`blocks`メソッドを追加

itagakisan committed:
受け取った文字数から最大幅を返すWidthクラスを追加 機能追加

[目的]
- 受け取った情報をフォーマットする際の最大幅を計算する

[変更内容]
- 07.ls_objectにおいて、width.rbを追加
- width.rbにWidthクラスを追加

itagakisan committed:
Widthクラスのテストを追加 テスト追加

[目的]
- Widthクラスの動作確認

[変更内容]
- LsTestクラスに`test_width`メソッドを追加
- Structを使用してファイル情報のモックを作成
- `update_max_widths`メソッドを複数回呼び出し、各カラムの最大幅が正しく更新されることをアサート

itagakisan committed:
Widthクラスに最大幅を返す機能を実装 機能追加

[目的]
- 各カラム（リンク数、オーナー、グループ、サイズ）の最大幅を計算・保持する責務をカプセル化

[変更内容]
- `initialize`で、各カラムの最大幅を保持するハッシュを初期化するよう実装
- ファイル情報を受け取り、各カラムの最大幅を更新する`update_max_widths`メソッドを実装

itagakisan committed:
コマンドラインオプションに基づいてファイルリストの取得とソート機能を実装 機能追加

[目的]
- オプションに対応したファイルリストの取得を行う

[変更内容]
- `LsOption`を使い、コマンドライン引数（-a, -r）を解釈する処理を実装
- `-a`オプションが指定された場合に隠しファイルを含めるため、`Dir.glob`に`File::FNM_DOTMATCH`フラグを渡す
- `-r`オプションが指定された場合にファイルリストを逆順にソートする処理を追加

itagakisan committed:
'-l'コマンドの有無によって取得する情報を変更する機能の実装 機能追加

[目的]
- ファイル名リストを、コマンドラインオプションに応じて、表示用のデータ構造（詳細情報ハッシュまたはファイル名文字列）に変換

[変更内容]
- ファイル名リストをループ処理し、ファイルごとに`FileInfo`オブジェクトを生成
- `-l`オプションが有効な場合、合計ブロック数を計算し、モード、リンク数、オーナーなどの詳細情報をハッシュとして収集
- `-l`オプションが無効な場合、ファイル名のみを収集
- 変換後の情報を`file_infos`配列に格納

itagakisan committed:
形式を整えて出力する機能の実装 機能追加

[目的]
- コマンドラインオプションに応じて、最終的な出力形式を整形し、表示する

[変更内容]
- 詳細フォーマット（-l）と標準の複数カラムフォーマットを整形するメソッド(`long_format_lines`、`format_filenames_table`)を追加
- ファイル情報をループ処理する際に、`Width`クラスを使って各カラムの最大表示幅を計算する処理を統合
- `-l`オプションの有無に基づき、適切なフォーマット処理を呼び出し、最終的な結果をコンソールに出力するロジックを実装

maedana commented on 07.ls_object/file_info.rb:
```diff
@@ -0,0 +1,50 @@
+      File.symlink?(filename) ? "#{filename} -> #{File.readlink(filename)}" : filename
+    else
+      filename
+    end
```
- このメソッドですが、elseのときは引数をそのまま返すだけですし、判定条件も引数依存しているだけで、このクラスのインスタンスメソッドとして実装する理由がないです。
- このクラス自体が基本的にlong_format用のものであり、long_format専用設計とするほうが責務として明確ではないですかね?
- 呼び出し元を見ても非long_formatのときは単にファイル名の配列を構築したいだけですし。

上記をもとに少し工夫すると引数不要のメソッドになるんじゃないかと思います。いかがでしょうか

gaki273 replied:
かしこまりました！
自分自身でも実装中にそのメソッドについて納得いっていませんでした。
ご指摘をもとに引数不要のメソッドになるように修正いたします。

maedana commented on 07.ls_object/ls.rb:
```diff
@@ -0,0 +1,72 @@
+      type: stats.file_type(filename, options)
+    }
+  else
+    stats.file_type(filename, options)
```
FileInfoのところで書きましたけど、この処理ってfilenames既にあるのにわざわざループして引数をそのまま返すメソッドで積め直してるだけなんですよね

gaki273 replied:
おっしゃる通りです。
修正いたします。

maedana commented on 07.ls_object/ls.rb:
```diff
@@ -0,0 +1,72 @@
+  max_widths = filenames.map(&:length).max
+  filename_lines = format_filenames_table(file_infos, max_widths, column)
+  filename_lines.each { |row| puts row.compact.join }
+end
```
メイン処理もっと短くしたいですねー。LsCommandみたいなクラスを作ってそっちに適切に移すことでメイン処理がもっとスッキリかけないですかね。

gaki273 replied:
かしこまりました！
クラスを作成し、メイン処理をより短くなるように修正いたします。

maedana commented on 07.ls_object/perm.rb:
```diff
@@ -0,0 +1,32 @@
+
+# frozen_string_literal: true
+
+class Perm
```
permって一般的な省略なんですかね? 変に省略せずに書いたほうがわかりやすいのでは?というようには感じましたがいかがでしょう?

gaki273 replied:
ご指摘のとおり、一般的に`perm`ですと、髪型のパーマの意味になりそうです。
`Permission`だと長いかな？と感じていましたが、変に省略してしまったせいで分かりにくくなってしまいました。
修正いたします。

maedana commented on 07.ls_object/perm.rb:
```diff
@@ -0,0 +1,32 @@
+    'link' => 'l',
+    'socket' => 's'
+  }.freeze
+  FILE_PERMISSION = {
```
細かいですが、定数定義の間は1行あけましょうー

gaki273 replied:
かしこまりました！
修正いたします。

maedana commented on 07.ls_object/perm.rb:
```diff
@@ -0,0 +1,32 @@
+    5 => 'r-x',
+    6 => 'rw-',
+    7 => 'rwx'
+  }.freeze
```
定数がどちらもクラス内でしか利用されてなさそうなので、privateにしたほうがいいのではないかなと思いました。やり方は調べてみてください :pray:

gaki273 replied:
かしこまりました。
`private`を用いるように修正いたします。

maedana commented on 07.ls_object/width.rb:
```diff
@@ -0,0 +1,23 @@
+    }
+  end
+
+  def update_max_widths(stats)
```
これだとnewしてからこのメソッドを引数付きで呼び出さないといけないので、利用側としては少し使いにくくないですか? `Width.new(stats).max_widths`で取得出来るほうがうれしいのではないかなと思いましたけどいかがですか? あとこのようにしておくと、`Width`の同じインスタンスであれば、`max_widths`は常に同じ結果を返してくるという安心感もあります。今の設計ですと、どこかで`update_max_widths`されて`max_widths`の内容が変わってるかもしれない、ということを考えないといけなくなります。

gaki273 replied:
ご指摘ありがとうございます。
確かに、現在の実装ですと
```ruby
w = Width.new
w.update_max_widths(stats)
w.max_widths
```
となり、冗長且つ、`update_max_widths`メソッドの使用箇所を考える必要もあり使いづらいですね。

ご提案のように
```ruby
w = Width.new(stats).max_widths
```
で取得するように修正いたします。

itagakisan committed:
`FileInfo`クラスのリファクタリング リファクタリング

[目的]
- `FileInfo`クラスの責務を、ファイルに関する情報を提供することに限定する

[変更内容]
- initialize時にファイル名をインスタンス変数として保持するように変更
- file_typeメソッドが引数を取らず、保持しているファイル名からファイル種別を返すように修正

itagakisan committed:
データ取得処理をリファクタリング リファクタリング

[目的]
- ファイル情報を取得するロジックを単純化

[変更内容]
- 条件分岐を削除
- 短い形式のフォーマットでは、ファイル名の配列を渡すように修正

itagakisan committed:
クラス名・ファイル名を変更、定数をプライベート化 リファクタリング

[目的]
- 何の役割を持つクラスなのかを明確化するため

[変更内容]
- クラス名を`Permission`に変更
- `perm.rb`を`permission.rb`にリネーム
- `file_info.rb`と`ls_test.rb`で、リネームしたファイルを正しく読み込むように`require_relative`のパスを修正
- `FileInfo`クラスの`file_mode`メソッド内で使用する変数名の変更と、クラス名を修正
- 定数を`private_constant`でプライベート化

itagakisan committed:
`Width`クラスのリファクタリング リファクタリング

[目的]
- 各列の最大幅を計算するロジックを単純化
- `Width`クラスをイミュータブル（不変）にし、インスタンス生成後に状態が変更されないようにすることで、コードの堅牢性を高める

[変更内容]
- `update_max_widths`メソッドを削除
- `initialize`メソッドが`FileInfo`オブジェクトの配列を受け取り、インスタンス化の際にすべての最大幅を計算するように変更

itagakisan committed:
テストコードを修正 修正

[目的]
- `Permission`, `FileInfo`, `Width`クラスのリファクタリングに伴い、古くなったテストコードを修正し、新しいクラスのインタフェースを正しく検証できるようにする

[変更内容]
- `test_perm`メソッドを`test_permission`にリネームし、テスト対象のクラス名を修正
- `test_file_info`メソッド内で、`file_type`メソッドの呼び出し方を修正
- `test_width`メソッドを全面的に書き直し、`Width`クラスがオブジェクトの配列で初期化される新しい設計に合わせてアサーションを修正

itagakisan committed:
`LsCommand`クラスを追加 機能追加

[目的]
- 主処理をクラス化することで、メインの処理を簡潔にする

[変更内容]
- コマンド実行のロジックを集約するための`LsCommand`クラスを追加

itagakisan committed:
`ls.rb`の主処理を`LsCommand`クラスへカプセル化 リファクタリング

[目的]
- コマンド実行のロジックを`LsCommand`クラスに集約する

[変更内容]
- `ls.rb`の処理を`LsCommand`クラスへ移行
- `ls.rb`は、`LsCommand`をインスタンス化して呼び出す1行のみに簡略化

maedana commented on 07.ls_object/ls_command.rb:
```diff
@@ -0,0 +1,61 @@
+    end
+  end
+
+  def load_filenames
```
クラス外から呼ばれないメソッドはカプセル化の観点からprivateにしましょうー

gaki273 replied:
かしこまりました！
クラス外から呼ばれないメソッドは`private`化して、修正いたします。

itagakisan committed:
`LsCommand`クラスと`LsOption`クラスの内部メソッドを`private`化 リファクタリング

[目的]
- 内部でのみ使用するメソッドを`private`にすることで、クラスの意図しない使われ方を防ぎ、より安全で堅牢な設計にする

[変更内容]
- `LsCommand`クラスにて、`load_filenames`, `format_filenames_table`, `long_format_lines` を`private`化
- `LsOption`クラスにて、`parse` を`private`化
